packages/@lwc/engine-core/src/framework/get-component-constructor.ts

```diff
@@ -6,21 +6,19 @@
  */
 
 import { isUndefined } from '@lwc/shared';
-import { getAssociatedVMIfPresent, LightningElement } from '@lwc/engine-core';
+import { LightningElement } from './base-lightning-element';
+import { getAssociatedVMIfPresent } from './vm';
 
 /**
  * EXPERIMENTAL: This function provides access to the component constructor, given an HTMLElement.
  * This API is subject to change or being removed.
  */
 export function getComponentConstructor(elm: HTMLElement): typeof LightningElement | null {
     let ctor: typeof LightningElement | null = null;
+    const vm = getAssociatedVMIfPresent(elm);
 
-    if (elm instanceof HTMLElement) {
```

Comment:
the check on whether or not `elm` is an HTMLElement is useless because this is a weakmap lookup that supports any kind of object as a key.

Reply:
> this is a weakmap lookup

Where is the weakmap? I don't quite understand this comment.

Reply:
> this is a weakmap lookup

Where is the weakmap? I don't quite understand this comment.

Reply:
May be to handle cases when `elm` is falsy? WeakMap.get(undefined) will throw.

`getComponentConstructor` is an exported function. The downstream user might not have typescript support and might invoke it with a undefined value.

Reply:
May be to handle cases when `elm` is falsy? WeakMap.get(undefined) will throw.

`getComponentConstructor` is an exported function. The downstream user might not have typescript support and might invoke it with a undefined value.

Reply:
it will not throw, but now I remember a case of a bad google map library that patched the weakmap prototype and throw on undefined value... I'm fine with adding back a condition that tests if it is an object, but not testing against htmlelement.

Reply:
OK, I found the weakmap you were talking about:

https://github.com/salesforce/lwc/blob/90b213b6c65834fe7dd33cf1f007a2314954a7f4/packages/%40lwc/engine-core/src/framework/vm.ts#L410-L411

If we're worried about passing in `undefined` to `WeakMap.prototype.get`, then yeah I think just checking for `undefined` is fine. Although maybe we should make it a dev-only check at the beginning of `getComponentConstructor` – it should never be `undefined` in prod, if the TS typechecking is working properly.

Reply:
since this is an internal experimental, only used by legacy locker, we should be fine adding the extra check.